File: 0207-course-schedule/0207-course-schedule.cpp
#include<vector>
#include<unordered_set>
using namespace std;

class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<unordered_set<int>> graph = make_graph(numCourses, prerequisites);
        vector<bool> onpath(numCourses, false), visited(numCourses, false);
        for(int i = 0; i < numCourses; i++) {
            if(!visited[i] && dfs_cycle(graph, i, onpath, visited))
                return false;
        }
        return true;
    }
private:
    vector<unordered_set<int>> make_graph(int numCourses, vector<vector<int>>& prerequisites){
        vector<unordered_set<int>> graph(numCourses);
        for(auto pre : prerequisites)
            graph[pre[1]].insert(pre[0]);
        return graph;
    }
    
    bool dfs_cycle(vector<unordered_set<int>>& graph, int node, vector<bool>& onpath, vector<bool>& visited){
        if(visited[node])
            return false;
        onpath[node] = visited[node] = true;
        for(int neigh : graph[node])
            if(onpath[neigh] || dfs_cycle(graph, neigh, onpath, visited))
                return true;
        return onpath[node] = false;
    }
};

File: 0207-course-schedule/0207-course-schedule.php
impl Solution {
     pub fn can_finish(num_courses: i32, prerequisites: Vec<Vec<i32>>) -> bool {
      let mut graph = vec![vec![]; num_courses as usize];
      let visited = &mut vec![0; num_courses as usize];
      for pre in prerequisites {
       let (x, y) = (pre[0], pre[1]);
       graph[x as usize].push(y);
      }
      for i in 0..num_courses {
       if !Solution::depth_first_search(graph.to_vec(), visited, i as usize) {
        return false;
       }
      }

      true
     }

     pub fn depth_first_search(graph: Vec<Vec<i32>>, visited: &mut Vec<i32>, i: usize) -> bool {
      if visited[i] == -1 {
       return false;
      } else if visited[i] == 1 {
       return true;
      }
      visited[i] = -1;
      for j in graph[i].to_vec() {
       if !Solution::depth_first_search(graph.to_vec(), visited, j as usize) {
        return false;
       }
      }
      visited[i] = 1;
      true
     }
}